### PR TITLE
feat: allow sharing OIDC state between clients over IPC MONGOSH-1386

### PR DESCRIPTION
**Pull request by addaleax**

### CODE REVIEW

File: src/ipc-rpc-state-share.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    this.server.listen({ path, exclusive: true });
+    await once(this.server, 'listening');
+    // The race condition in applying chmod after listening is just inherent
+    // to the Node.js API for sockets. We said we'd set the file mode to 0o600
+    // in the design and it probably doesn't hurt to do this either way.
+    await fs.chmod(path, 0o600);
+
+    // In practice, mongosh and Compass may not always know when a connection
+    // state fully closes, so we're adding a process 'exit' listener that cleans
+    // up the file if it hasn't bee closed so far.
```
Comment (alenakhineika) [inline src/ipc-rpc-state-share.ts]:
```suggestion
    // up the file if it hasn't been closed so far.
```

File: src/ipc-rpc-state-share.spec.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      refresh.resolves({ accessToken: 'ref-accesstoken' });
+    });
+
+    afterEach(async function() {
+      await server.close();
+    });
+
+    it('can be used to share OIDC state', async function() {
+      server = await StateShareServer.create(fakeState);
+      client = new StateShareClient(server.handle);
```
Comment (alenakhineika) [inline src/ipc-rpc-state-share.spec.ts:127]: Is it something that we will do in vscode? or DataService will wrap this API? I am now thinking, in vscode, the connection controller uses DataService, but the language server uses CliServiceProvider to connect. We might need then to plan work to replace CliServiceProvider connection in LS with DataServie.

File: src/ipc-rpc-state-share.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    this.server.close();
+    await once(this.server, 'close');
+    this.cleanupCallback?.();
+    this.cleanupCallback = undefined;
+  }
+
+  protected abstract handleRpc(content: Record<string, unknown>): Promise<Record<string, unknown>>;
+}
+
+// Client correspondong to RpcServer instances.
```
Comment (alenakhineika) [inline src/ipc-rpc-state-share.ts]:
```suggestion
// Client corresponding to RpcServer instances.
```

File: src/ipc-rpc-state-share.spec.ts (hunk 2 above)
Comment (addaleax) [reply src/ipc-rpc-state-share.spec.ts:127]: > Is it something that we will do in vscode? 

Use this feature? Yes. Create these objects here manually? No, they’re implementation detials.

> or DataService will wrap this API? I am now thinking, in vscode, the connection controller uses DataService, but the language server uses CliServiceProvider to connect. We might need then to plan work to replace CliServiceProvider connection in LS with DataServie.

I don’t see an issue here. Both use devtools-connect to create the connection, so they should be compatible.

File: src/ipc-rpc-state-share.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+export class RpcClient {
+  private hmacKey: Buffer
```
Comment (alenakhineika) [inline src/ipc-rpc-state-share.ts]:
```suggestion
  private hmacKey: Buffer;
```

File: src/ipc-rpc-state-share.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  private serverAddress: string
```
Comment (alenakhineika) [inline src/ipc-rpc-state-share.ts]:
```suggestion
  private serverAddress: string;
```

File: src/ipc-rpc-state-share.spec.ts (hunk 2 above)
Comment (addaleax) [reply src/ipc-rpc-state-share.spec.ts:127]: We’re actually already relying on the fact that the options are compatible between them, I think: https://github.com/mongodb-js/vscode/blob/4d0c41d8d289b2ff66a5ec5c621f74d06a59177d/src/connectionController.ts#L833-L837

File: src/ipc-rpc-state-share.ts (hunk 4 above)
Comment (addaleax) [reply src/ipc-rpc-state-share.ts]: Yeah … prettier would be nice to have here :)

**Review by alenakhineika — APPROVED** [review]
Amazing work! Was very interesting to go through the code with all the comments, super helpful! 🔥

**Pull request merged**